proof-systems/src/darlin/marlin/src/constraints/test.rs

```diff
@@ -0,0 +1,206 @@
+            assert!(Marlin::<G, PC>::verify(&index_vk, &pc_vk, &correct_inputs, &proof).unwrap());
+
+            let mut cs = ConstraintSystem::<G::BaseField>::new(SynthesisMode::Debug);
+
+            let verifier_key_gadget = VerifierKeyGadget::<G, PC, PCG>::alloc_input(
+                cs.ns(|| "alloc verifier key"),
+                || Ok(index_vk.clone()),
+            )
+            .unwrap();
+
+            let pc_verifier_key_gadget =
+                PCG::VerifierKey::alloc(cs.ns(|| "alloc pc verifier key"), || Ok(pc_vk.clone()))
+                    .unwrap();
+
+            let proof_gadget =
+                ProofGadget::<G, PC, PCG>::alloc_input(cs.ns(|| "alloc proof"), || {
+                    Ok(proof.clone())
+                })
+                .unwrap();
+
+            let public_inputs_gadget =
+                PublicInputsGadget::<G>::alloc_input(cs.ns(|| "alloc public inputs"), || {
+                    Ok(correct_inputs)
+                })
+                .unwrap();
+
+            MarlinVerifierGadget::<G, GG, PC, PCG>::verify(
+                cs.ns(|| "proof verification"),
+                &pc_verifier_key_gadget,
+                &verifier_key_gadget,
+                public_inputs_gadget,
+                &proof_gadget,
+            )
+            .unwrap();
+
+            if !cs.is_satisfied() {
+                println!(
+                    "constraint {} is unsatisfied",
+                    cs.which_is_unsatisfied().unwrap()
+                )
+            }
+
+            assert!(cs.is_satisfied());
+
+            // Fail verification
+            let wrong_inputs = vec![a, a];
+            assert!(!Marlin::<G, PC>::verify(&index_vk, &pc_vk, &wrong_inputs, &proof).unwrap());
+
+            let mut cs = ConstraintSystem::<G::BaseField>::new(SynthesisMode::Debug);
+
+            let verifier_key_gadget = VerifierKeyGadget::<G, PC, PCG>::alloc_input(
+                cs.ns(|| "alloc verifier key"),
+                || Ok(index_vk.clone()),
+            )
+            .unwrap();
+
+            let pc_verifier_key_gadget =
+                PCG::VerifierKey::alloc(cs.ns(|| "alloc pc verifier key"), || Ok(pc_vk.clone()))
+                    .unwrap();
+
+            let proof_gadget =
+                ProofGadget::<G, PC, PCG>::alloc_input(cs.ns(|| "alloc proof"), || {
+                    Ok(proof.clone())
+                })
+                .unwrap();
+
+            let public_inputs_gadget =
+                PublicInputsGadget::<G>::alloc_input(cs.ns(|| "alloc public inputs"), || {
+                    Ok(wrong_inputs)
+                })
+                .unwrap();
+
+            MarlinVerifierGadget::<G, GG, PC, PCG>::verify(
+                cs.ns(|| "proof verification"),
+                &pc_verifier_key_gadget,
+                &verifier_key_gadget,
+                public_inputs_gadget,
+                &proof_gadget,
+            )
+            .unwrap();
+
+            assert!(!cs.is_satisfied());
```

Comment:
Like the primitive, are we able to trigger failure due to sumchecks and failure due to opening proof verification, and assert explicitly that the corresponding specific constraints fail ?

Reply:
Done. Actually, I've split the verifier gadget into IOP check and succinct opening proof check, as done in primitive.